torchtune/modules/tokenizer.py

```diff
@@ -89,3 +118,59 @@ def decode(self, ids: List[int]) -> str:
             str: The decoded text.
         """
         return self.spm_model.decode(ids)
+
+    def tokenize_messages(
+        self, messages: List[Message], max_seq_len: Optional[int] = None
+    ) -> Tuple[List[int], List[bool]]:
+        start_of_turn = True
+        end_of_turn = False
+        tokenized_messages = []
+        mask = []
+        for message in messages:
+            # If assistant message, this is the end of a turn
+            end_of_turn = message.role == "assistant"
```

Comment:
I'm grossly misunderstanding, but how are both start_of_turn and end_of_turn both True here?

Reply:
They shouldn't be. End of turn is just on an assistant message. We set that at the beginning of the iteration, then at the end of the iteration, we set end of turn to False and start of turn to True, since the next iteration will be the start of a turn. Lmk if I'm misunderstanding your question here though

Reply:
start of turn and end of turn can both be true here if the list of messages the user passes in consists of only a single assistant message. This would not be a valid dialogue, but the BOS and EOS should still be appended correctly. I wonder if we should validate that the dialogue is well-formed here or outside this method

Reply:
Yeah good point. Imo it's not up to the tokenizer to decide this. Let's consider adding a separate utility e.g. `validate_messages` that can be called from the dataset class or elsewhere as a follow-up.